src/qt3utils/datagenerators/plescanner.py

```diff
@@ -0,0 +1,266 @@
+import logging
+import time
+import tkinter as tk
+from typing import Type
+
+import numpy as np
+
+from qt3utils.applications.controllers.nidaqedgecounter import QT3ScanNIDAQEdgeCounterController
+from qt3utils.applications.controllers.wavemeter_controller import WavemeterController
+from qt3utils.applications.controllers.lockin_controller import Lockin
+from qt3utils.nidaq.customcontrollers import VControl
+
+logger = logging.getLogger(__name__)
+
+class PleScanner:
+    """
+    Base class with function methods which any PLE experiment will need to operate
+    Override this class for specific types of PLE experiments
+    """
+
+    def __init__(self, readers, wavelength_controller):
+        """
+        self.running: boolean true when scan thread is running
+        self.current_frame: current pixel (or time) number in sweep
+        self._start: minimum allowed point for hardware controlling the laser wavelength
+        self._end: maximum allowed point for hardware controlling the laser wavelength
+        self.tmax: maximum number of pixels (or time) for sweep
+        self._step_size: size between steps controlling wavelength as it is swept
+        self.wavelength_controller: the wavelength controller class instance
+        """
+        self.running = False
+        self.current_frame = 0
+        self._start = float(wavelength_controller.minimum_allowed_position)
+        self._end= float(wavelength_controller.maximum_allowed_position)
+        self.tmax = None
+        self._step_size = None
+        self.wavelength_controller = wavelength_controller
+        self.readers = readers
+        self.outputs = []
+        self.rate_counters = []
+        self.scanned_control = []
+        self.scan_mode = ""
+        self._discrete_batch_size = 1
+        if isinstance(self.wavelength_controller, VControl):
+            self.axis_name = "Voltage"
+        else:
+            self.axis_name = "Wavelength"
+
+    def stop(self) -> None:
+        """
+        stop running scan
+        """
+        self.running = False
+        for rate_counter in self.rate_counters:
+            rate_counter.stop()
```

Comment:
When does the main qt3ple stop? it seems like it stops after an entire single axis scan is finished. This may incorrectly terminate devices. It is better to have a stop method that tells the loop to stop, and another post-stop method that closes the devices.

Reply:
This stop command just changes the "running" flag to false, which is queried via the "still_scanning" function. If "still_scanning" returns false, the background thread sweeping the wavelength and taking measurements will stop sweeping and exit. The connection to the devices should not change until you configure with a new YAML file or close the application. 
Is this working as you expected, or do you still want a change?

Reply:
What you describe is what we want. We discssused this in person today. You mentioned putting everything in a loop under the close method.